src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
Moreover, this code is present not for correctness of the generated code, but for its quality.

---
In reply to: [100643714](https://github.com/dotnet/roslyn/pull/16928#discussion_r100643714) [](ancestors = 100643714,100629991,100629224,100627591,100616402,100602791,100601157,100591895)